projects/feray/feray/service.py

```diff
@@ -0,0 +1,270 @@
+    ):
+        self.registry = registry
+        self._fs = feature_store
+        self._ms = metadata_store
+        self.hasher = hasher
+        self.data_version_fn = data_version_fn
+        self.row_id_fn = row_id_fn
+        self.logger = logging.getLogger(__name__)
+
+    def materialize(
+        self, asset_keys: list[S] | None = None, force_recompute: bool = False
+    ):
+        """Materializes a set of assets, respecting dependencies and caching."""
+        target_assets = asset_keys or list(self.registry.definitions.keys())
+        execution_order = self.registry.resolve_execution_order(target_assets)
+        self.logger.info(
+            "\n--- Starting Batch Materialization (Force=%s) ---", force_recompute
+        )
+        for asset_key in execution_order:
+            self._materialize_one(asset_key, force_recompute=force_recompute)
+
+    def delete_entity_data(
+        self, *, start_asset: S, entity_key_values: dict[str, Any], cascade: bool = True
+    ):
+        """Deletes data for a given entity, with optional cascading to downstream assets."""
+        self.logger.info(
+            "\n--- Deleting data for entity %s starting from '%s' ---",
+            entity_key_values,
+            start_asset,
+        )
+
+        # --- THE FIX: Create a canonical hash string that matches the one in polars_add_row_id_fn ---
+        definition = self.registry.definitions[start_asset]
+        try:
+            # Build the string by joining the entity key values in their defined order.
+            canonical_string = "".join(
+                [str(entity_key_values[k]) for k in definition.entity_keys]
+            )
+        except KeyError as e:
+            raise ValueError(
+                f"Missing entity key {e} in provided values for asset '{start_asset}'."
+            ) from e
+
+        row_id_to_delete = RowId(hashlib.sha256(canonical_string.encode()).hexdigest())
+
+        assets_to_process = self._resolve_cascading_assets(start_asset, cascade)
+        self.logger.info("Deletion will apply to assets: %s", assets_to_process)
+
+        for asset_key in assets_to_process:
+            self._fs.delete_ids(asset_key=asset_key, row_ids=[row_id_to_delete])
+
+    def prune_asset_versions(
+        self,
+        *,
+        asset_key: S,
+        keep_last_n: int | None = None,
+        max_age: timedelta | None = None,
+    ):
+        """Prunes old data and metadata versions for a given asset based on policies."""
+        self.logger.info("\n--- Pruning versions for asset '%s' ---", asset_key)
+        all_prov = self._ms.all_provenance(asset_key=asset_key)
+        if not all_prov:
+            return
+
+        provs_to_delete: list[Provenance] = []
+        if keep_last_n is not None and len(all_prov) > keep_last_n:
+            provs_to_delete = all_prov[:-keep_last_n]
+
+        if max_age is not None:
+            self.logger.warning("WARNING: Pruning by max_age is not fully implemented.")
```

Comment:
Should this just raise a `NotImplementedError` or perhaps just be removed from the interface until it's  implemented? Right now it seems like the `max_age` parameter just doesn't do anything, which might be confusing for users. Kind of another explicit vs. implicit thing